app/models/vm_scan.rb

```diff
@@ -384,105 +238,36 @@ def process_data(*args)
     # got data to process
   end
 
-  def delete_snapshot(mor, vm = nil)
-    vm ||= VmOrTemplate.find(target_id)
-    if mor
-      begin
-        if vm.ext_management_system
-          if options[:snapshot] == :smartProxy
-            log_end_user_event_message(vm)
-            delete_snapshot_by_description(mor, vm)
-          else
-            user_event = end_user_event_message(vm)
-            if vm.kind_of?(ManageIQ::Providers::Openstack::CloudManager::Vm)
-              vm.ext_management_system.vm_delete_evm_snapshot(vm, mor)
-            elsif vm.kind_of?(ManageIQ::Providers::Microsoft::InfraManager::Vm) ||
-                  (vm.kind_of?(ManageIQ::Providers::Azure::CloudManager::Vm) && vm.require_snapshot_for_scan?)
-              vm.ext_management_system.vm_delete_evm_snapshot(vm, :snMor => mor)
-            else
-              vm.ext_management_system.vm_remove_snapshot(vm, :snMor => mor, :user_event => user_event)
-            end
-          end
-        else
-          raise _("No Providers available to delete snapshot")
-        end
-      rescue => err
-        _log.error(err.message)
-        _log.log_backtrace(err, :debug)
-      end
-    else
-      log_end_user_event_message(vm)
-    end
-  end
-
-  def delete_snapshot_by_description(mor, vm)
-    if mor
-      ems_type = 'host'
-      options[:ems_list] = vm.ems_host_list
-      miqVimHost = options[:ems_list][ems_type]
-
-      miqVim = nil
-      # Make sure we were given a host to connect to and have a non-nil encrypted password
-      if miqVimHost && !miqVimHost[:password].nil?
-        server = miqVimHost[:hostname] || miqVimHost[:ipaddress]
-        begin
-          password_decrypt = ManageIQ::Password.decrypt(miqVimHost[:password])
-          if MiqServer.use_broker_for_embedded_proxy?(ems_type)
-            $vim_broker_client ||= MiqVimBroker.new(:client, MiqVimBrokerWorker.drb_port)
-            miqVim = $vim_broker_client.getMiqVim(server, miqVimHost[:username], password_decrypt)
-          else
-            require 'VMwareWebService/MiqVim'
-            miqVim = MiqVim.new(server, miqVimHost[:username], password_decrypt)
-          end
-
-          vimVm = miqVim.getVimVm(vm.path)
-          vimVm.removeSnapshotByDescription(mor, true) unless vimVm.nil?
-        ensure
-          vimVm.release if vimVm rescue nil
-          miqVim.disconnect unless miqVim.nil?
-        end
-      end
-    end
-  end
-
   def user_event_message(vm, verb)
     "EVM SmartState Analysis #{verb} for VM [#{vm.name}]"
   end
 
-  def start_user_event_message(vm)
+  def start_user_event_message(vm = nil)
+    vm ||= VmOrTemplate.find(target_id)
```

Comment:
This `VmOrTemplate.find(target_id)` pattern seems all over, so I wonder if it's better to just have a `vm` method that memoizes this.

Reply:
I agree - but would it make sense to do it in a future refactoring PR?

Reply:
On the other hand, this code is all over the subclasses too.  So it will be easier to refactor now.  Given that the name of the class is `VmScan`, having a memoized `vm` instance method seems quite appropriate.

Changing to WIP in order to get this refactoring in.

Reply:
Refactoring with new `vm` method is in.